File: qpi-driver/js/src/events.ts
/**
 * Event envelope shared with QPI-UI over NNG (RFC 0001 §4, §6).
 *
 * A driver speaks the same typed events QPI-UI understands: it handles the
 * events QPI-UI sends it and emits events of its own. Every event travels in
 * one envelope whose payload shape depends on its type and is validated by
 * whoever handles it.
 */

import { randomBytes } from "node:crypto";

/**
 * The fixed set of event types a QPI-UI version understands. Maintainers grow
 * the framework by adding new types over releases.
 */
export enum EventType {
  JobDispatch = "JobDispatch",
  JobResult = "JobResult",
  CryostatReading = "CryostatReading",
}

/** The wire shape of an envelope, as sent over NNG. */
export interface EventWire {
  id: string;
  driver: string;
  type: EventType;
  ts: string;
  payload: Record<string, unknown>;
}

/** Optional envelope fields; normally left to the SDK to fill in. */
export interface EventInit {
  driver?: string;
  id?: string;
  ts?: string;
}

/** A single typed message exchanged with QPI-UI in either direction. */
export class Event {
  readonly type: EventType;
  readonly payload: Record<string, unknown>;
  /** Identifier of the driver this event belongs to; stamped on emit. */
  driver: string;
  readonly id: string;
  readonly ts: string;

  constructor(
    type: EventType,
    payload: Record<string, unknown> = {},
    init: EventInit = {},
  ) {
    this.type = type;
    this.payload = payload;
    this.driver = init.driver ?? "";
    this.id = init.id ?? newEventId();
    this.ts = init.ts ?? nowTimestamp();
  }

  /** Return the envelope as a plain object matching the wire shape. */
  toWire(): EventWire {
    return {
      id: this.id,
      driver: this.driver,
      type: this.type,
      ts: this.ts,
      payload: this.payload,
    };
  }

  /** Serialise the envelope to a JSON string for sending over NNG. */
  toJSON(): string {
    return JSON.stringify(this.toWire());
  }

  /** Build an event from a JSON string or bytes received over NNG. */
  static fromJSON(raw: string | Buffer): Event {
    const text = typeof raw === "string" ? raw : raw.toString("utf8");
    const data = JSON.parse(text) as Partial<EventWire>;
    if (!data.type) {
      throw new Error("event envelope is missing a type");
    }
    return new Event(data.type, data.payload ?? {}, {
      driver: data.driver ?? "",
      id: data.id ?? "",
      ts: data.ts ?? "",
    });
  }
}

function newEventId(): string {
  return "evt_" + randomBytes(12).toString("hex");
}

function nowTimestamp(): string {
  return new Date().toISOString().replace(/(\.\d{3})\d*Z$/, "$1Z");
}

File: qpi-driver/go/events.go
// Package qpidriver is the Go SDK for building QPI drivers (RFC 0001 §4).
//
// A driver is an external process that exchanges typed events with QPI-UI: it
// handles the events QPI-UI sends it and emits events of its own. Embed
// [Base] in a struct, implement [Handler.HandleEvent] to act on each inbound
// event (switching on its type), and call [Base.Emit] to send an event upward.
// [Base.Every] runs a callback on a timer, for drivers that report on their
// own schedule rather than in reply to a dispatch. [Run] performs the
// handshake, opens the transport, and blocks until the process is signalled to
// stop.
//
// It mirrors the Python SDK (qpi-driver/py/qpi_driver) and the TypeScript SDK
// (qpi-driver/js): the same envelope, the same drivers/connect handshake, and
// TLS with the pinned root CA.
package qpidriver

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"time"
)

// EventType is one of the fixed event types a QPI-UI version understands.
// Maintainers grow the framework by adding new types over releases.
type EventType string

const (
	// JobDispatch is sent by QPI-UI to a QPU driver: run this job.
	JobDispatch EventType = "JobDispatch"
	// JobResult is emitted by a QPU driver once a job finishes.
	JobResult EventType = "JobResult"
	// CryostatReading is emitted by a monitor driver on its own schedule.
	CryostatReading EventType = "CryostatReading"
)

// Event is a single typed message exchanged with QPI-UI in either direction.
// The Payload shape depends on Type and is validated by whoever handles the
// event. The wire form is the shared envelope {id, driver, type, ts, payload}
// (RFC 0001 §6).
type Event struct {
	// ID uniquely identifies this envelope; generated when empty.
	ID string `json:"id"`
	// Driver identifies the driver this event belongs to.
	Driver string `json:"driver"`
	// Type determines the payload shape.
	Type EventType `json:"type"`
	// Ts is the creation time, ISO-8601 UTC with millisecond precision;
	// generated when empty.
	Ts string `json:"ts"`
	// Payload is the type-specific body.
	Payload map[string]any `json:"payload"`
}

// NewEvent builds an event of the given type for a driver, filling in a fresh
// ID and timestamp. It is the idiomatic way to construct an event to Emit.
func NewEvent(eventType EventType, driver string, payload map[string]any) Event {
	if payload == nil {
		payload = map[string]any{}
	}
	return Event{
		ID:      newEventID(),
		Driver:  driver,
		Type:    eventType,
		Ts:      nowTimestamp(),
		Payload: payload,
	}
}

// MarshalJSON serialises the envelope, filling in ID, Ts, and a non-nil
// Payload so a zero-valued Event still produces a well-formed envelope.
func (e Event) MarshalJSON() ([]byte, error) {
	if e.ID == "" {
		e.ID = newEventID()
	}
	if e.Ts == "" {
		e.Ts = nowTimestamp()
	}
	if e.Payload == nil {
		e.Payload = map[string]any{}
	}
	type wire Event
	return json.Marshal(wire(e))
}

// decodeEvent parses a received wire message into an Event.
func decodeEvent(raw []byte) (Event, error) {
	var e Event
	if err := json.Unmarshal(raw, &e); err != nil {
		return Event{}, err
	}
	if e.Payload == nil {
		e.Payload = map[string]any{}
	}
	return e, nil
}

func newEventID() string {
	buf := make([]byte, 12)
	_, _ = rand.Read(buf)
	return "evt_" + hex.EncodeToString(buf)
}

func nowTimestamp() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05.000") + "Z"
}
